### PR TITLE
refactor(mysql): use MySQLdb native driver, drop ibis dependency

### PR DESCRIPTION
**Pull request by goldmedal**
## Summary

The mysql / doris connector now uses `mysqlclient` (MySQLdb) directly instead of going through ibis. The `mysql` extra installs `mysqlclient` only.

## Changes

- `connector/mysql.py`: native MySQLdb cursor + field-type-code → Arrow type mapping covering integer (signed/unsigned variants), decimal, float, char/varchar/text, json, blob, datetime/timestamp, date, time, bit. SSL kwargs are derived from `MySqlConnectionInfo`.
- `model/data_source.py`: `get_mysql_connection` / `get_doris_connection` now return a `MySQLdb.Connection`; SSL context construction moved into the connector module.
- `pyproject.toml`: `mysql` extra → `mysqlclient>=2.2` (drops `ibis-framework[mysql]`).

## Test plan

- [x] `tests/connectors/test_mysql_connector.py` covers 17 MySQL field-type cases via a testcontainer mysql (tinyint signed/unsigned, smallint, int, bigint, decimal, float/double, char/varchar/text, json, blob, varbinary, datetime/timestamp, date, time, year, bit, enum/set, null handling, limit, dry_run, empty result).
- [x] Existing `tests/connectors/test_mysql.py` TPCH suite continues to pass (9 cases) on the new native driver.
- [x] Unit tests green (`just test-unit`).
- [x] `just lint` clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Refactor**
  * Switched MySQL/Doris connectors to a native MySQL driver with improved connection handling and robust conversion to Arrow types (numeric, decimals, binary, JSON, and temporal edge cases).
* **Chore**
  * Simplified optional MySQL dependencies by removing the ibis extra.
* **Tests**
  * Added extensive integration and unit tests covering type conversions, LIMIT handling, dry-run, temporal boundaries, and concurrency-safe helper initialization.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Canner/WrenAI/pull/2273)

### CODE REVIEW

File: core/wren/src/wren/connector/mysql.py [hunk 1]
```diff
@@ -1,57 +1,378 @@
...
-class DorisConnector(IbisConnector):
+        self.connection = DataSource.mysql.get_connection(connection_info)
+        # Append ANSI_QUOTES to the server-configured sql_mode so identifiers
+        # quoted as "name" (the MDL convention) are accepted. CONCAT preserves
+        # the server defaults (ONLY_FULL_GROUP_BY, STRICT_TRANS_TABLES, …) —
+        # overwriting them would let queries silently behave differently than
+        # in the user's own MySQL session.
+        with closing(self.connection.cursor()) as cursor:
+            cursor.execute("SET sql_mode=CONCAT(@@sql_mode, ',ANSI_QUOTES')")
+        self._closed = False
```
[the remark this answers is not in the document]
Comment (goldmedal) [reply core/wren/src/wren/connector/mysql.py]: Addressed in 8c6e3cd5.

File: core/wren/src/wren/connector/mysql.py [hunk 2]
```diff
@@ -1,57 +1,378 @@
...
+        if limit is not None:
+            sql = f"SELECT * FROM ({sql}) AS _sub LIMIT {limit}"
+        with closing(self.connection.cursor()) as cursor:
+            cursor.execute(sql)
+            return _build_mysql_arrow_table(cursor)
+
+    def dry_run(self, sql: str) -> None:
+        with closing(self.connection.cursor()) as cursor:
+            cursor.execute(f"SELECT * FROM ({sql}) AS _sub LIMIT 0")
+            cursor.fetchall()
```
[the remark this answers is not in the document]
Comment (goldmedal) [reply core/wren/src/wren/connector/mysql.py:101]: Addressed in 8c6e3cd5 via `_coerce_limit` (int validation). Also switched to LIMIT-append via `_apply_limit` so duplicate-column queries (joins exposing the same column twice) no longer trip `ER_DUP_FIELDNAME`.

File: core/wren/src/wren/connector/mysql.py [hunk 3]
```diff
@@ -1,57 +1,378 @@
...
+    base_map: dict[str, pa.DataType] = {
+        "TINY": pa.int8(),
+        "SHORT": pa.int16(),
+        "LONG": pa.int32(),
+        "INT24": pa.int32(),
+        "LONGLONG": pa.int64(),
+        "FLOAT": pa.float32(),
+        "DOUBLE": pa.float64(),
+        "DECIMAL": pa.decimal128(38, 9),
+        "NEWDECIMAL": pa.decimal128(38, 9),
```
[the remark this answers is not in the document]
Comment (goldmedal) [reply core/wren/src/wren/connector/mysql.py]: Addressed in 8c6e3cd5 via `_arrow_decimal_from_mysql_field` — precision and scale are now derived from `cursor.description` instead of being hard-coded.

File: core/wren/src/wren/connector/mysql.py [hunk 4]
```diff
@@ -1,57 +1,378 @@
...
+        "ENUM": pa.string(),
+        "SET": pa.string(),
+        "TINY_BLOB": pa.binary(),
+        "MEDIUM_BLOB": pa.binary(),
+        "LONG_BLOB": pa.binary(),
+        "BLOB": pa.binary(),
+        "JSON": pa.string(),
+        "DATE": pa.date32(),
+        "NEWDATE": pa.date32(),
+        "TIME": pa.time64("us"),
```
[the remark this answers is not in the document]
Comment (goldmedal) [reply core/wren/src/wren/connector/mysql.py]: Addressed in 0c503b22 — `FIELD_TYPE.TIME` now maps to `pa.duration("us")` and the value path converts MySQLdb's `datetime.timedelta` to signed microseconds explicitly. Regression tests cover `-100:00:00`, `0:00:00`, `838:59:59` (max) and `-838:59:59` (min).

File: core/wren/src/wren/connector/mysql.py [hunk 5]
```diff
@@ -1,57 +1,378 @@
...
+
+        host = parsed.hostname or "127.0.0.1"
+        if host == "localhost":
+            host = "127.0.0.1"
+
+        out: dict = {
+            "host": host,
+            "port": int(parsed.port) if parsed.port else 3306,
+            "user": parsed.username,
+            "passwd": unquote_plus(parsed.password) if parsed.password else "",
```
[the remark this answers is not in the document]
Comment (goldmedal) [reply core/wren/src/wren/connector/mysql.py:479]: `unquote_plus` here is consistent with how URL-derived passwords are handled elsewhere in the repo (`+` decoded as space). In practice userinfo callers pass unencoded values, so the round-trip works. Happy to revisit if a real encoding mismatch surfaces.

**Comment by goldmedal** [comment]
Thanks for the review — quick disposition of the review-body items not posted as inline threads:

- **`get_connection` return-type widening** — addressed in ffce4ff9 by introducing a `ConnectionHandle` union so `MySQLdb.Connection` is part of the declared return contract alongside `BaseBackend`.
- **`@classmethod` → `@staticmethod` on the URL/connection helpers** — leaving as-is. They're module-private, behaviour is identical, and the unused `cls` is cosmetic. Happy to revisit alongside a broader cleanup of the other native builders.
- **`DorisConnector` duplicates `MySqlConnector.__init__`** — leaving as-is. Doris intentionally diverges (no ANSI_QUOTES init, different connection routing) so the duplication is load-bearing — extracting hooks before a second Doris-specific code path lands would be premature.

And the one new commit on top:

- **TIME → `pa.duration("us")`** — addressed in 0c503b22 (the previously-reported inline thread). Regression tests exercise `-100:00:00`, `0:00:00`, `838:59:59` (max) and `-838:59:59` (min) so the full MySQL TIME range round-trips.

File: core/wren/src/wren/connector/mysql.py [hunk 6]
```diff
@@ -1,57 +1,518 @@
...
+        if code is not None:
+            _MYSQL_BLOB_CODES.add(code)
+    for n in ("STRING", "VAR_STRING", "VARCHAR"):
+        code = getattr(FT, n, None)
+        if code is not None:
+            _MYSQL_STRING_CODES.add(code)
+    for n in ("DECIMAL", "NEWDECIMAL"):
+        code = getattr(FT, n, None)
+        if code is not None:
+            _MYSQL_DECIMAL_CODES.add(code)
```
[the remark this answers is not in the document]
Comment (goldmedal) [reply core/wren/src/wren/connector/mysql.py]: Addressed in 1983232a.

File: core/wren/src/wren/connector/mysql.py [hunk 7]
```diff
@@ -1,57 +1,518 @@
...
+    if not rows:
+        arrays = [pa.array([], type=field.type) for field in schema]
+    else:
+        arrays = [
+            _build_mysql_column([row[i] for row in rows], schema.field(i).type)
+            for i in range(len(fields))
+        ]
+    return pa.table(
+        dict(zip([f.name for f in fields], arrays, strict=False)),
+        schema=schema,
```
[the remark this answers is not in the document]
Comment (goldmedal) [reply core/wren/src/wren/connector/mysql.py]: Addressed in 2b2e9c0f.

**Comment by goldmedal** [comment]
Addressed CodeRabbit round-3 nit by pinning the testcontainer image tag from `mysql:8` to `mysql:8.0.36` in both `tests/connectors/test_mysql_connector.py` and `tests/connectors/test_mysql.py` in a387af90.

**Comment by goldmedal** [comment]
Superseded by #2313 — these seven native-driver refactors were consolidated into a single feature branch to resolve shared-file conflicts (data_source.py, pyproject.toml, uv.lock, factory.py, etc.) once instead of seven times.
